super_editor/lib/src/core/document_composer.dart

```
@@ -29,25 +32,75 @@ class DocumentComposer with ChangeNotifier {
   @override
   void dispose() {
     _preferences.dispose();
+    selectionNotifier.removeListener(_onSelectionChangedBySelectionNotifier);
     super.dispose();
   }
 
   /// Returns the current [DocumentSelection] for a [Document].
-  DocumentSelection? get selection => selectionNotifier.value;
+  DocumentSelection? get selection => _latestSelectionChange.selection;
 
-  /// Sets the current [selection] for a [Document].
+  /// Sets the current [selection] for a [Document] using [SelectionReason.userInteraction] as the reason.
   set selection(DocumentSelection? newSelection) {
     if (newSelection != selectionNotifier.value) {
       selectionNotifier.value = newSelection;
       notifyListeners();
     }
   }
 
+  /// Sets the current [selection] for a [Document].
+  ///
+  /// [reason] represents what caused the selection change to happen.
+  void setSelection(DocumentSelection? newSelection, [Object reason = SelectionReason.userInteraction]) {
+    _latestSelectionChange = DocumentSelectionChange(
+      selection: newSelection,
+      reason: reason,
+    );
+    _streamController.sink.add(_latestSelectionChange);
+
+    if (_updatingSelection) {
+      // The selection was changed by the selectionNotifier.
+      return;
+    }
+
+    _updatingSelection = true;
+    try {
```

Why do we need a try/catch here?

---

I added this to avoid some listener throwing an exception and preventing `_updatingSelection` to be reset.

---

If something is throwing exceptions then we should fix the exceptions, rather than silently hide them.

---

I mean exceptions in user code which could listen for selection changes. BTW, this is a try/finally so it won't hide any exception.

---

What leads you to believe that there will be exceptions in user code? I don't think we wrap callbacks with try/catch anywhere else. I don't think I've seen any special reason to do that here...

---

Ok, I think we can just remove it.